### PR TITLE
[NNC] Lock thread/block computation

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #73800

Differential Revision: [D34647281](https://our.internmc.facebook.com/intern/diff/D34647281)

Since this is invoked multi-threaded, we have to lock the access to evaluators. This could cause an error if multiple threads are running the same kernel and computing the block extent in parallel.

### CODE REVIEW

File: test/cpp/tensorexpr/test_dynamic_shapes.cpp [hunk 1]
```diff
@@ -687,4 +687,4 @@ TEST(DynamicShapes, MultiThreadedExecution) {
 }
 
 } // namespace jit
-} // namespace torch
+} // namespace torch
```
Comment (navahgar) [inline test/cpp/tensorexpr/test_dynamic_shapes.cpp]: unrelated edit

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34647281).

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34647281).

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34647281).

**Review by navahgar — APPROVED** [review]
LGTM

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34647281).
